File: Velix/app/lib/services/media_service.dart
import 'dart:convert';
import 'dart:io';
import 'package:crypto/crypto.dart';
import 'package:file_picker/file_picker.dart';
import 'package:flutter/foundation.dart';
import 'package:flutter_dotenv/flutter_dotenv.dart';
import 'package:http/http.dart' as http;
import 'package:image_picker/image_picker.dart';
import 'package:path_provider/path_provider.dart';

class MediaResult {
  final String url;
  final String publicId;
  MediaResult({required this.url, required this.publicId});
}

class PickedFile {
  final File file;
  final String name;
  PickedFile({required this.file, required this.name});
}

class MediaService {
  final _picker = ImagePicker();

  Future<File?> pickImage() async {
    final picked = await _picker.pickImage(
        source: ImageSource.gallery, imageQuality: 80);
    if (picked == null) return null;
    return File(picked.path);
  }

  Future<File?> captureImage() async {
    final picked = await _picker.pickImage(
        source: ImageSource.camera, imageQuality: 85);
    if (picked == null) return null;
    return File(picked.path);
  }

  Future<File?> pickVideo() async {
    final picked = await _picker.pickVideo(source: ImageSource.gallery);
    if (picked == null) return null;
    return File(picked.path);
  }

  Future<File?> captureVideo() async {
    final picked = await _picker.pickVideo(source: ImageSource.camera);
    if (picked == null) return null;
    return File(picked.path);
  }

  Future<PickedFile?> pickFile() async {
    // withData:true ensures bytes are loaded even under Android scoped storage
    final result =
        await FilePicker.platform.pickFiles(withData: true);
    if (result == null || result.files.isEmpty) return null;
    final f = result.files.first;

    File file;
    if (f.bytes != null) {
      // Write bytes to a temp file so we have a normal File path for upload
      final tmp = await getTemporaryDirectory();
      final tmpFile = File('${tmp.path}/${f.name}');
      await tmpFile.writeAsBytes(f.bytes!);
      file = tmpFile;
    } else if (f.path != null) {
      file = File(f.path!);
    } else {
      return null;
    }

    return PickedFile(file: file, name: f.name);
  }

  // Signed upload using API Key + Secret — no upload preset required.
  // resourceType: 'image' | 'video' | 'raw'
  Future<(MediaResult?, String?)> upload(
      String uid, File file, String resourceType) async {
    try {
      final cloudName = dotenv.env['CLOUDINARY_CLOUD_NAME'] ?? '';
      final apiKey = dotenv.env['CLOUDINARY_API_KEY'] ?? '';
      final apiSecret = dotenv.env['CLOUDINARY_API_SECRET'] ?? '';

      if (cloudName.isEmpty) return (null, '缺少 CLOUDINARY_CLOUD_NAME');
      if (apiKey.isEmpty) return (null, '缺少 CLOUDINARY_API_KEY');
      if (apiSecret.isEmpty) return (null, '缺少 CLOUDINARY_API_SECRET');

      final timestamp = DateTime.now().millisecondsSinceEpoch ~/ 1000;
      final folder = 'pchats/$uid';

      // Params must be sorted alphabetically before signing
      final toSign = 'folder=$folder&timestamp=$timestamp';
      final signature =
          sha256.convert(utf8.encode(toSign + apiSecret)).toString();

      final uri = Uri.parse(
          'https://api.cloudinary.com/v1_1/$cloudName/$resourceType/upload');

      final request = http.MultipartRequest('POST', uri)
        ..fields['api_key'] = apiKey
        ..fields['timestamp'] = timestamp.toString()
        ..fields['signature'] = signature
        ..fields['folder'] = folder
        ..files.add(await http.MultipartFile.fromPath('file', file.path));

      final streamed =
          await request.send().timeout(const Duration(seconds: 60));
      final response = await http.Response.fromStream(streamed);

      debugPrint('[MediaService] status=${response.statusCode}');
      debugPrint('[MediaService] body=${response.body}');

      if (response.statusCode != 200) {
        String errMsg = 'HTTP ${response.statusCode}';
        try {
          final body = jsonDecode(response.body) as Map<String, dynamic>;
          final msg = body['error']?['message'] as String?;
          if (msg != null && msg.isNotEmpty) errMsg = msg;
        } catch (_) {}
        return (null, errMsg);
      }

      final data = jsonDecode(response.body) as Map<String, dynamic>;
      return (
        MediaResult(
          url: data['secure_url'] as String,
          publicId: data['public_id'] as String,
        ),
        null,
      );
    } catch (e) {
      debugPrint('[MediaService] upload exception: $e');
      return (null, e.toString());
    }
  }

  Future<void> delete(String publicId, String resourceType) async {
    try {
      final cloudName = dotenv.env['CLOUDINARY_CLOUD_NAME'] ?? '';
      final apiKey = dotenv.env['CLOUDINARY_API_KEY'] ?? '';
      final apiSecret = dotenv.env['CLOUDINARY_API_SECRET'] ?? '';

      final timestamp = DateTime.now().millisecondsSinceEpoch ~/ 1000;
      final toSign = 'public_id=$publicId&timestamp=$timestamp';
      final signature =
          sha256.convert(utf8.encode(toSign + apiSecret)).toString();

      await http.post(
        Uri.parse(
            'https://api.cloudinary.com/v1_1/$cloudName/$resourceType/destroy'),
        body: {
          'public_id': publicId,
          'api_key': apiKey,
          'timestamp': timestamp.toString(),
          'signature': signature,
          'signature_algorithm': 'sha256',
        },
      );
    } catch (_) {}
  }
}

File: P-Chats/frontend/src/services/mediaService.ts
const CLOUD_NAME = 'dgopeoebw'
const API_KEY = '549262742261114'
const API_SECRET = '4lp1DL1oYdS1Y8Jt-RkVYu5UZZc'

async function sha256hex(s: string): Promise<string> {
  const buf = await crypto.subtle.digest('SHA-256', new TextEncoder().encode(s))
  return Array.from(new Uint8Array(buf))
    .map(b => b.toString(16).padStart(2, '0'))
    .join('')
}

export interface MediaResult {
  url: string
  publicId: string
}

class MediaService {
  async upload(uid: string, file: File, resourceType: string): Promise<MediaResult | null> {
    const timestamp = Math.floor(Date.now() / 1000)
    const folder = `pchats/${uid}`
    const toSign = `folder=${folder}&timestamp=${timestamp}${API_SECRET}`
    const signature = await sha256hex(toSign)

    const form = new FormData()
    form.append('file', file)
    form.append('api_key', API_KEY)
    form.append('timestamp', String(timestamp))
    form.append('signature', signature)
    form.append('folder', folder)

    const res = await fetch(
      `https://api.cloudinary.com/v1_1/${CLOUD_NAME}/${resourceType}/upload`,
      { method: 'POST', body: form },
    )
    if (!res.ok) return null
    const data = await res.json() as { secure_url: string; public_id: string }
    return { url: data.secure_url, publicId: data.public_id }
  }

  async delete(publicId: string, resourceType: string): Promise<void> {
    const timestamp = Math.floor(Date.now() / 1000)
    const toSign = `public_id=${publicId}&timestamp=${timestamp}${API_SECRET}`
    const signature = await sha256hex(toSign)

    const form = new FormData()
    form.append('public_id', publicId)
    form.append('api_key', API_KEY)
    form.append('timestamp', String(timestamp))
    form.append('signature', signature)
    form.append('signature_algorithm', 'sha256')

    await fetch(
      `https://api.cloudinary.com/v1_1/${CLOUD_NAME}/${resourceType}/destroy`,
      { method: 'POST', body: form },
    ).catch(() => {})
  }
}

export const mediaService = new MediaService()
